Improve general states configurability

gergondet opened the pull request:
- Allow to specify a set of constraints that are added by the state and removed at completion
- Allow to specify a set of tasks that are added by the state and removed at completion
- ^^ `MetaTasks` now uses the capability above and simply load the completion criteria
- RemovePostureTask now removes all active FSM posture tasks or can be specified to remove some of the active posture tasks

Along the way:
- add a test to check that all common state options work correctly
- introduce `oneOf` support for properties in schema doc generation, I will look into supporting this better for some tasks specifications (for now it won't affect the panel)
- fix `mc_rtc::Configuration::empty()` to return false if the configuration holds a non-null value
- fix a long standing bug where loading a nested configuration into another configuration would load the entire document under some circumstances

arntanguy reviewed (changes requested):
Great suggestions!

A few points to sum up my comments above:
- We no longer get a message when tasks/constraints are added/removed from the solver. I think that's a good default display to keep.
- The constraints name is ignored
- It'd be nice to be able to link directly to a JSON schema category, e.g https://jrl-umi3218.github.io/mc_rtc/json-full.html#ConstraintSet-objects and expand the category, and get a detailed description for that category instead of the blank element we currently get (same when we click on it).

gergondet commented on doc/_data/schemas/common/State.json:
```diff
@@ -17,6 +17,16 @@
+    {
+      "oneOf": [
+        { "type": "boolean", "default": false, "description": "If true, remove all posture tasks at the state's start." },
+        { "type": "object", "description": "Keys are robots' names, value is a boolean indicating whether the posture taks is removed." }
```
Great point. Done in [`2fde530` (#80)](https://github.com/jrl-umi3218/mc_rtc/pull/80/commits/2fde530c8be7646d29c521c7076887188920f6b3)

I've opted to:
1. Keep `RemovePostureTask` working for backward-compatibility
2. Keep a singular form on `DisablePostureTask` as it makes it less likely there will be an issue due to the "double" change

gergondet commented on doc/_data/schemas/common/State.json:
```diff
@@ -17,6 +17,16 @@
+    "constraints": { "type": "object",
+                     "description": "Constraints that will be added when the state starts and removed when the state stops.<br/>Keys are used to identify the constraints in the description." },
+    "tasks": { "type": "object",
+               "description": "Tasks that will be added when the state starts and removed when the state stops.<br/>Keys are the name of the tasks and values are MetaTask objects." }
```
Yes.

Implemented in [`202b4d5` (#80)](https://github.com/jrl-umi3218/mc_rtc/pull/80/commits/202b4d543f451a801477bd91f76c2cd79883d5ad) for the landing page and [`14ab446` (#80)](https://github.com/jrl-umi3218/mc_rtc/pull/80/commits/14ab4468cf65926c75c46783cc9b3aec0dac9a19) for linking from the state description

The link are a little simpler. You just have to omit any object name to land on the category description

gergondet commented on src/mc_control/fsm/State.cpp:
```diff
@@ -105,18 +120,71 @@ void State::start_(Controller & ctl)
+    for(const auto & c : constraints)
+    {
+      constraints_.push_back(mc_solver::ConstraintSetLoader::load(ctl.solver(), c.second));
+      ctl.solver().addConstraintSet(*constraints_.back());
```
This is the case in mc_rtc 2.0 so I'd rather keep it as-is for now than introduce a new feature.

gergondet commented on src/mc_control/fsm/State.cpp:
```diff
@@ -105,18 +120,71 @@ void State::start_(Controller & ctl)
+      }
+      tasks_.push_back(mc_tasks::MetaTaskLoader::load(ctl.solver(), tConfig));
+      ctl.solver().addTask(tasks_.back());
+    }
```
Tasks messages (added/removed) are handled by `mc_solver::QPSolver` not by the state so this doesn't change anything.

gergondet commented on tests/controllers/TestFSMStateOptions.in.yaml:
```diff
@@ -0,0 +1,90 @@
+            body2: R_HIP_Y_S
+            iDist: 0.05
+            sDist: 0.025
+            damping: 0
```
Done in [`820d12f` (#80)](https://github.com/jrl-umi3218/mc_rtc/pull/80/commits/820d12f5aff503b395bff79bbd3af0d40c465121)

gergondet commented:
Thanks for the review. I've addressed your comments and implemented your suggestions for the most part.

I left the `ConstraintSet` changes aside for now as this is "fixed" in 2.0 and not super easy to do in this versions (e.g. to get meaningful information from the Tasks constraint)
